driver/src/orderbook/batched_auction_data_reader.rs

```diff
@@ -0,0 +1,336 @@
+use crate::contracts::stablex_auction_element::{StableXAuctionElement, AUCTION_ELEMENT_WIDTH};
+use crate::models::{AccountState, BatchInformation, Order};
+use std::collections::HashMap;
+use web3::types::{H160, U256};
+
+/// Handles reading of auction data that has been encoded with the smart
+/// contract's `encodeAuctionElement` function.
+pub struct BatchedAuctionDataReader {
+    /// The account state resulting from unfiltered handled orders.
+    account_state: AccountState,
+    /// All unfiltered orders in the order they were received.
+    orders: Vec<Order>,
+    /// Used when reading data from the smart contract in batches.
+    pagination: Pagination,
+    /// The index of batch whose orders we are filtering for.
+    index: U256,
+    /// The total number of orders per user.
+    user_order_counts: HashMap<H160, usize>,
+}
+
+/// Data for the next call to the smart contract's `getEncodedUsersPaginated`
+/// function (which should be named `getEncodedOrdersPaginated`).
+pub struct Pagination {
+    /// The user of the last received order or H160::zero when no order has been
+    /// received.
+    pub previous_page_user: H160,
+    /// The number of received orders for `previous_page_user`.
+    pub previous_page_user_offset: usize,
+}
+
+/// The resulting order data of a finishes BatchedAuctionDataReader.
+pub struct Done {
+    /// The account state resulting from unfiltered handled orders.
+    pub account_state: AccountState,
+    /// All unfiltered orders in the order they were received.
+    pub orders: Vec<Order>,
+}
+
+impl BatchedAuctionDataReader {
+    /// Create a new BatchedAuctionDataReader.
+    pub fn new(index: U256) -> BatchedAuctionDataReader {
+        BatchedAuctionDataReader {
+            account_state: AccountState::default(),
+            orders: Vec::new(),
+            pagination: Pagination {
+                previous_page_user: H160::zero(),
+                previous_page_user_offset: 0,
+            },
+            index,
+            user_order_counts: HashMap::new(),
+        }
+    }
+
+    /// The pagination data used when reading data from the smart contract in
+    /// batches.
+    pub fn pagination(&self) -> &Pagination {
+        &self.pagination
+    }
+
+    /// Signal that no more data will be read and return the result.
+    pub fn done(self) -> Done {
```

Comment:
What's the point of defining the Done struct? Shouldn't we just return a tuple? I'd also prefer if this method was called `get_auction_data` or something a little bit more explicit. There is nothing enforcing the no more data will be read.

Reply:
It takes `self` by value which consumes self which means it cannot be used anymore after the call.
Struct vs Tuple is a minor difference. With structs you can name the values with tuples you have only types. I prefer structs.

Reply:
Nice, I didn’t spot the consumption of self (maybe this can be pointed out in the name or doc string). 

I still think the method name is too generic and would argue that the definition of the Done struct for the purpose of packing the return type into a single value adds unnecessary indirection and cognitive overhead than just returning account state and orders without defining the extra type. If you feel strongly about keeping the struct I’d ask to give it a more descriptive name.

Reply:
I like the `done` method, its like finalizing a builder. I do agree with @fleupold that maybe it can return a tuple instead.